smartsim/_core/utils/timings.py

```diff
@@ -0,0 +1,134 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import time
+import typing as t
+from collections import OrderedDict
+
+import numpy as np
+
+from ...log import get_logger
+
+logger = get_logger("PerfTimer")
+
+
+class PerfTimer:
```

Comment:
Consider implementing as a context manager to avoid having to manually mark the end of a section:

For example:
```
class PerfTimer:
    def __init__(self) -> None:
        self._start_time = 0
        self._end_time = 0
        self._name = ""

    @property
    def name(self) -> str:
        return self._name
    
    @property
    def duration(self) -> t.Optional[int]:
        if self._end_time == 0:
            return None        
        return self._end_time - self._start_time

    def __enter__(self, name: str, verbose: bool = True):
        self._name = name
        self._verbose = verbose

        self._start_time = time.perf_counter_ns()
        if self._verbose:
            print(f"Starting {self.name}")

    def __exit__(self, exc_type, exc_value, exc_tb):
        self._end_time = time.perf_counter_ns()
        if self._verbose:
            print(f"Completed {self.name} in {self.duration}")

```

would let us avoid the manual stop calls, but nesting stinks:

```

with PerfTimer("receive"):
    response = MessageHandler.deserialize_response(resp)

    with PerfTimer("deserialize_response"):
        data_blob: bytes = from_recvh.recv_bytes(timeout=None)
```

a version that uses context vars would make it a bit nicer by avoiding the nested context managers:

```
class PerfTimer:
    KEY: str = "perf_timer"

    def __init__(self, name: str, verbose: bool = True) -> None:
        self._verbose = verbose
        self._ctx = contextvars.ContextVar(PerfTimer.KEY)

    def __enter__(self):
        self._reset_token = self._ctx.set(time.perf_counter_ns())

    def __exit__(self, exc_type, exc_value, exc_tb):
        self._duration = time.perf_counter_ns() - int(self._ctx.get())
        self._ctx.reset(self._reset_token)
```

Reply:
Can we put this on another ticket? I am even fine removing the `PerfTimer` class before merging, actually, it is a "good-enough" thing which helps with getting some quick data, but I did not really design it as a production tool